File: sdks/python/src/raven/resources/tokens.py
"""The core of this SDK (Phase 10 spec Section 15): mint a short-lived RTC
token server-side, then hand it to your frontend — never mint a token in the
browser, and never store a minted token any longer than you need to forward
it. Every token is short-lived by design; there is no way to request a
permanent one (see ``expires_in``).
"""

from __future__ import annotations

from typing import Any, cast

from .._http import RavenHttpClient
from .._async_http import AsyncRavenHttpClient
from .._types import CreateTokenParams, IssuedToken


def _token_body(params: CreateTokenParams) -> dict[str, Any]:
    return {
        "participantIdentity": params.identity,
        "permissions": params.permissions.to_json() if params.permissions else None,
        "ttlSeconds": params.expires_in,
        "metadata": params.metadata,
    }


class TokensResource:
    def __init__(self, http: RavenHttpClient) -> None:
        self._http = http

    def create(self, params: CreateTokenParams) -> IssuedToken:
        result = self._http.request(f"/v1/rooms/{params.room}/rtc-tokens", method="POST", body=_token_body(params))
        return cast(IssuedToken, result)


class AsyncTokensResource:
    def __init__(self, http: AsyncRavenHttpClient) -> None:
        self._http = http

    async def create(self, params: CreateTokenParams) -> IssuedToken:
        result = await self._http.request(f"/v1/rooms/{params.room}/rtc-tokens", method="POST", body=_token_body(params))
        return cast(IssuedToken, result)

File: packages/server-sdk/src/resources/tokens.ts
import type { RavenHttpClient } from '../http-client';
import type { CreateTokenParams, IssuedToken } from '../types';

/**
 * The heart of this SDK (Phase 10 spec §15). Mint a short-lived RTC token
 * server-side, then hand it to your frontend.
 *
 * Never mint a token in the browser, and never hold on to a minted one any
 * longer than it takes to forward it. Every token is short-lived by design,
 * and there's no way to ask for a permanent one (see `expiresIn`).
 */
export class TokensResource {
  constructor(private readonly http: RavenHttpClient) {}

  create(params: CreateTokenParams): Promise<IssuedToken> {
    const { room, identity, permissions, expiresIn, metadata } = params;
    return this.http.request<IssuedToken>(`/v1/rooms/${room}/rtc-tokens`, {
      method: 'POST',
      body: {
        participantIdentity: identity,
        permissions,
        ttlSeconds: expiresIn,
        metadata,
      },
    });
  }
}
